File: src/memory/tokenBudget.ts
/**
 * Token budget management for LLM context window allocation.
 *
 * Ensures that context compression, memory recall, and active task
 * all stay within the model's context window limits.
 */

export interface TokenBudgetConfig {
  total?: number;
  systemPrompt?: number;
  compressionRatio?: number;
  memoryRatio?: number;
  taskRatio?: number;
  bufferRatio?: number;
}

export class TokenBudget {
  readonly total: number;
  readonly systemPrompt: number;
  readonly compressionRatio: number;
  readonly memoryRatio: number;
  readonly taskRatio: number;
  readonly bufferRatio: number;

  constructor(config: TokenBudgetConfig = {}) {
    this.total = config.total ?? 128_000;
    this.systemPrompt = config.systemPrompt ?? 2_000;
    this.compressionRatio = config.compressionRatio ?? 0.4;
    this.memoryRatio = config.memoryRatio ?? 0.1;
    this.taskRatio = config.taskRatio ?? 0.4;
    this.bufferRatio = config.bufferRatio ?? 0.1;
  }

  /** Total tokens available after system prompt. */
  get available(): number {
    return this.total - this.systemPrompt;
  }

  /** Max tokens allocated for compressed history. */
  get compressionBudget(): number {
    return Math.floor(this.available * this.compressionRatio);
  }

  /** Max tokens allocated for memory recall. */
  get memoryBudget(): number {
    return Math.floor(this.available * this.memoryRatio);
  }

  /** Max tokens allocated for active task context. */
  get taskBudget(): number {
    return Math.floor(this.available * this.taskRatio);
  }

  /** Remaining buffer for unexpected growth. */
  get bufferBudget(): number {
    return Math.floor(this.available * this.bufferRatio);
  }

  /**
   * Calculate remaining budget for memory after compression.
   * If compression used less than its budget, the surplus goes to memory.
   */
  remainingForMemory(compressedTokens: number): number {
    const compressionSurplus = this.compressionBudget - compressedTokens;
    const availableForMemory = this.memoryBudget + Math.max(0, compressionSurplus);
    return Math.max(0, availableForMemory);
  }
}

File: src/memory/token_budget.py
"""Token budget management for LLM context window allocation."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class TokenBudget:
    """Manages token allocation across different context components.

    Ensures that context compression, memory recall, and active task
    all stay within the model's context window limits.
    """

    total: int = 128_000
    system_prompt: int = 2_000
    compression_ratio: float = 0.4
    memory_ratio: float = 0.1
    task_ratio: float = 0.4
    buffer_ratio: float = 0.1

    @property
    def available(self) -> int:
        """Total tokens available after system prompt."""
        return self.total - self.system_prompt

    @property
    def compression_budget(self) -> int:
        """Max tokens allocated for compressed history."""
        return int(self.available * self.compression_ratio)

    @property
    def memory_budget(self) -> int:
        """Max tokens allocated for memory recall."""
        return int(self.available * self.memory_ratio)

    @property
    def task_budget(self) -> int:
        """Max tokens allocated for active task context."""
        return int(self.available * self.task_ratio)

    @property
    def buffer_budget(self) -> int:
        """Remaining buffer for unexpected growth."""
        return int(self.available * self.buffer_ratio)

    def remaining_for_memory(self, compressed_tokens: int) -> int:
        """Calculate remaining budget for memory after compression.

        If compression used less than its budget, the surplus goes to memory.
        If compression exceeded its budget, memory budget is reduced.
        """
        compression_surplus = self.compression_budget - compressed_tokens
        available_for_memory = self.memory_budget + max(0, compression_surplus)
        return max(0, available_for_memory)
